File: src/brand-identity/services/credits_service.py
"""
Entitlement gate and credit reserve/commit/refund for Brand Identity.

Two independent checks before any generation runs:
  1. Entitlement — does this subscription plan include Brand Identity at all?
     Explorer and Student do not. This is a static check; no DB call needed.
  2. Credits — does the user have enough shared credits remaining this cycle?
     This is an atomic UPDATE against the main GMBTE platform DB.

Reserve happens BEFORE the generation pipeline is queued.
Commit happens in the worker on JobStatus.DONE.
Refund happens in the worker on any failure path.

Every operation appends a row to `ai_credit_transactions` so all AI services
share one audit trail. This is append-only, not a mutable-status row per
reservation — reserve/commit/refund each write their own row.

SCHEMA: built against
  user_credits(user_id, credits_balance, credits_reset_at)
  ai_credit_transactions(id, user_id, service, amount, type, reference_id, created_at)

Reconcile against the real schema by editing this file and core/credits_db.py only.
"""
import logging
import uuid
from sqlalchemy import text
from sqlalchemy.ext.asyncio import AsyncSession

logger = logging.getLogger(__name__)

SERVICE_NAME = "brand_identity"

TIER_RANK_SQL = text("""
    SELECT tier FROM subscriptions WHERE user_id = :user_id
""")

# Matches the rank order of the SubscriptionTier enum in prisma/schema.prisma.
# Values are the actual Postgres enum labels Prisma writes (EXPLORER, STUDENT,
# ...), not the old founder_workspace/founder_pro placeholder strings —
# nothing in the codebase ever produced those; they were dead vocabulary
# invented before the real subscriptions table existed. Users with no row
# yet default to EXPLORER (matches CreditsService.getOrCreateSubscription on
# the NestJS side).
TIER_RANK = {
    "EXPLORER": 0,
    "STUDENT": 1,
    "PROFESSIONAL": 2,
    "FOUNDER": 3,
    "EXECUTIVE": 4,
    "TEAM": 5,
    "ENTERPRISE": 6,
}

MIN_TIER = "EXECUTIVE"


class EntitlementError(Exception):
    """Raised when a plan doesn't include Brand Identity."""


class InsufficientCreditsError(Exception):
    """Raised when the user's credit balance is too low."""


class CreditsDbError(Exception):
    """Raised when the credits DB operation itself fails."""


async def check_entitlement(db: AsyncSession, user_id: str) -> None:
    """
    Gate 1 — is this user's real subscription tier at or above MIN_TIER?
    One cheap read against the shared subscriptions table (same DB as the
    credit ledger). Raises EntitlementError if not.
    """
    result = await db.execute(TIER_RANK_SQL, {"user_id": user_id})
    row = result.fetchone()
    tier = row[0] if row else "EXPLORER"

    if TIER_RANK.get(tier, 0) < TIER_RANK[MIN_TIER]:
        raise EntitlementError(
            f"Your current plan ({tier}) does not include "
            f"Brand Identity. Please upgrade to {MIN_TIER.title()} or above."
        )


async def reserve_credits(
    db: AsyncSession,
    user_id: str,
    cost: int,
    asset_id: str,
) -> None:
    """
    Atomically deduct `cost` credits from the user's balance.
    Fails CLOSED — any DB error raises CreditsDbError rather than granting
    a free generation.

    Writes a 'reserved' row to credit_transactions for the audit trail.
    """
    try:
        result = await db.execute(
            text(
                """
                UPDATE user_credits
                SET credits_balance = credits_balance - :cost
                WHERE user_id = :user_id
                  AND credits_balance >= :cost
                RETURNING credits_balance
                """
            ),
            {"user_id": user_id, "cost": cost},
        )
        row = result.fetchone()
        if row is None:
            raise InsufficientCreditsError(
                f"Insufficient credits. {cost} credits are required for Brand Identity generation."
            )

        await db.execute(
            text(
                """
                INSERT INTO ai_credit_transactions
                    (id, user_id, service, amount, type, reference_id, created_at)
                VALUES
                    (:id, :user_id, :service, :amount, 'reserve', :ref, NOW())
                """
            ),
            {
                "id": str(uuid.uuid4()),
                "user_id": user_id,
                "service": SERVICE_NAME,
                "amount": -cost,
                "ref": asset_id,
            },
        )
        await db.commit()
        logger.info(
            "Credits reserved: user=%s cost=%d asset=%s remaining=%d",
            user_id, cost, asset_id, row[0],
        )

    except (EntitlementError, InsufficientCreditsError):
        raise
    except Exception as exc:
        logger.exception("Credits DB error during reserve for user=%s asset=%s", user_id, asset_id)
        raise CreditsDbError("Credits system unavailable. Please try again.") from exc


async def commit_credits(
    db: AsyncSession,
    user_id: str,
    cost: int,
    asset_id: str,
) -> None:
    """
    Mark a previously reserved credit transaction as 'committed' (generation succeeded).
    This is a ledger update only — the balance was already deducted at reserve time.
    """
    try:
        await db.execute(
            text(
                """
                INSERT INTO ai_credit_transactions
                    (id, user_id, service, amount, type, reference_id, created_at)
                VALUES
                    (:id, :user_id, :service, 0, 'commit', :ref, NOW())
                """
            ),
            {
                "id": str(uuid.uuid4()),
                "user_id": user_id,
                "service": SERVICE_NAME,
                "ref": asset_id,
            },
        )
        await db.commit()
        logger.info("Credits committed: user=%s cost=%d asset=%s", user_id, cost, asset_id)
    except Exception as exc:
        # Non-fatal — generation succeeded, only the ledger update failed.
        # Log for manual reconciliation; do not fail the job.
        logger.error(
            "Failed to commit credit transaction for user=%s asset=%s: %s",
            user_id, asset_id, exc,
        )


async def refund_credits(
    db: AsyncSession,
    user_id: str,
    cost: int,
    asset_id: str,
) -> None:
    """
    Refund `cost` credits when generation fails after a successful reserve.
    Writes a 'refunded' row to the ledger for the audit trail.
    """
    try:
        await db.execute(
            text(
                """
                UPDATE user_credits
                SET credits_balance = credits_balance + :cost
                WHERE user_id = :user_id
                """
            ),
            {"user_id": user_id, "cost": cost},
        )
        await db.execute(
            text(
                """
                INSERT INTO ai_credit_transactions
                    (id, user_id, service, amount, type, reference_id, created_at)
                VALUES
                    (:id, :user_id, :service, :cost, 'refund', :ref, NOW())
                """
            ),
            {
                "id": str(uuid.uuid4()),
                "user_id": user_id,
                "service": SERVICE_NAME,
                "cost": cost,
                "ref": asset_id,
            },
        )
        await db.commit()
        logger.info("Credits refunded: user=%s cost=%d asset=%s", user_id, cost, asset_id)
    except Exception as exc:
        # Refund failure means the user was charged for a failed generation.
        # Log at ERROR level so this can be caught and corrected manually.
        logger.error(
            "CRITICAL: Failed to refund credits for user=%s asset=%s cost=%d: %s",
            user_id, asset_id, cost, exc,
        )

File: src/credits/credits.service.ts
import { Injectable, Logger } from '@nestjs/common';
import { SubscriptionTier, SubscriptionStatus } from '@prisma/client';
import { randomUUID } from 'crypto';
import { PrismaService } from '../prisma/prisma.service';

/** Matches the rank order of the SubscriptionTier enum in schema.prisma.
 *  "Min tier" checks (per the tool registry in the pricing doc) are a floor,
 *  not an explicit per-tool allow-list, so this stays a single ordered
 *  array rather than one ENTITLED_PLANS set per tool. */
const TIER_RANK: Record<SubscriptionTier, number> = {
  [SubscriptionTier.EXPLORER]: 0,
  [SubscriptionTier.STUDENT]: 1,
  [SubscriptionTier.PROFESSIONAL]: 2,
  [SubscriptionTier.FOUNDER]: 3,
  [SubscriptionTier.EXECUTIVE]: 4,
  [SubscriptionTier.TEAM]: 5,
  [SubscriptionTier.ENTERPRISE]: 6,
};

/** Default monthly allowance per tier — used only when a user has no
 *  Subscription row yet (implicit Explorer) or when granting a fresh one.
 *  Kept here rather than duplicated at every call site. Team's 1,500 is a
 *  shared-pool number and doesn't actually fit this per-user model yet —
 *  see the flag in gmbte-pricing memory; Team accounts default to Executive's
 *  per-seat number until that's designed properly. */
const DEFAULT_MONTHLY_ALLOWANCE: Record<SubscriptionTier, number> = {
  [SubscriptionTier.EXPLORER]: 10,
  [SubscriptionTier.STUDENT]: 50,
  [SubscriptionTier.PROFESSIONAL]: 150,
  [SubscriptionTier.FOUNDER]: 250,
  [SubscriptionTier.EXECUTIVE]: 1000,
  [SubscriptionTier.TEAM]: 1000, // placeholder — shared pool not yet modeled
  [SubscriptionTier.ENTERPRISE]: 1000, // placeholder — custom per contract
};

export class EntitlementError extends Error {}
export class InsufficientCreditsError extends Error {}

export interface CreditReservation {
  referenceId: string;
  userId: string;
  service: string;
  cost: number;
}

@Injectable()
export class CreditsService {
  private readonly logger = new Logger(CreditsService.name);

  constructor(private readonly prisma: PrismaService) {}

  /** For the frontend's proactive balance check — e.g. Academy course cards
   *  showing "Enroll (80 credits)" before the user taps anything, or the
   *  pay-confirmation screen the doc calls for. Creates the implicit
   *  EXPLORER subscription/balance on first touch, same as reserve() does,
   *  so a brand-new user gets a real number back instead of an error. */
  async getBalance(userId: string): Promise<{
    balance: number;
    tier: SubscriptionTier;
    monthlyAllowance: number;
  }> {
    const sub = await this.getOrCreateSubscription(userId);
    const credits = await this.getOrCreateCredits(userId, sub.monthlyAllowance);
    return {
      balance: credits.creditsBalance,
      tier: sub.tier,
      monthlyAllowance: sub.monthlyAllowance,
    };
  }

  /** Earning credits, not spending them — e.g. the Green Impact eco-point
   *  exchange. No entitlement/tier check (anyone can earn), no reserve step
   *  (there's nothing to fail-closed against), just a balance increment plus
   *  an append-only ledger row so it shows in the same audit trail as every
   *  spend. Returns the new balance. */
  async grant(
    userId: string,
    service: string,
    amount: number,
  ): Promise<number> {
    const sub = await this.getOrCreateSubscription(userId);
    await this.getOrCreateCredits(userId, sub.monthlyAllowance);

    const updated = await this.prisma.userCredits.update({
      where: { userId },
      data: { creditsBalance: { increment: amount } },
    });

    await this.prisma.aiCreditTransaction.create({
      data: {
        userId,
        service,
        referenceId: randomUUID(),
        amount,
        type: 'grant',
        balanceAfter: updated.creditsBalance,
      },
    });

    return updated.creditsBalance;
  }

  /** Gate 1 — static-ish check (one cheap read, no external call). Every
   *  user implicitly has an EXPLORER/MANUAL subscription if no row exists
   *  yet, matching "Explorer: free, no signup step beyond auth" from the
   *  pricing tiers. */
  async checkEntitlement(
    userId: string,
    minTier: SubscriptionTier,
  ): Promise<void> {
    const sub = await this.getOrCreateSubscription(userId);
    if (TIER_RANK[sub.tier] < TIER_RANK[minTier]) {
      throw new EntitlementError(
        `This feature requires ${minTier} tier or above (current: ${sub.tier}).`,
      );
    }
  }

  private async getOrCreateSubscription(userId: string) {
    const existing = await this.prisma.subscription.findUnique({
      where: { userId },
    });
    if (existing) return existing;

    return this.prisma.subscription.create({
      data: {
        userId,
        tier: SubscriptionTier.EXPLORER,
        status: SubscriptionStatus.MANUAL,
        monthlyAllowance: DEFAULT_MONTHLY_ALLOWANCE[SubscriptionTier.EXPLORER],
      },
    });
  }

  private async getOrCreateCredits(userId: string, allowance: number) {
    const existing = await this.prisma.userCredits.findUnique({
      where: { userId },
    });
    if (existing) return existing;

    return this.prisma.userCredits.create({
      data: { userId, creditsBalance: allowance, creditsResetAt: null },
    });
  }

  /** Gate 2 — atomic conditional decrement, fails closed on any DB error
   *  (mirrors the Python services' fail-closed reserve). Writes an append-only
   *  'reserve' row to ai_credit_transactions. */
  async reserve(
    userId: string,
    service: string,
    cost: number,
  ): Promise<CreditReservation> {
    const sub = await this.getOrCreateSubscription(userId);
    await this.getOrCreateCredits(userId, sub.monthlyAllowance);

    const referenceId = randomUUID();

    const result = await this.prisma.userCredits.updateMany({
      where: { userId, creditsBalance: { gte: cost } },
      data: { creditsBalance: { decrement: cost } },
    });

    if (result.count === 0) {
      throw new InsufficientCreditsError(
        `Insufficient credits: ${cost} required for ${service}.`,
      );
    }

    const updated = await this.prisma.userCredits.findUnique({
      where: { userId },
    });
    await this.prisma.aiCreditTransaction.create({
      data: {
        userId,
        service,
        referenceId,
        amount: -cost,
        type: 'reserve',
        balanceAfter: updated?.creditsBalance ?? null,
      },
    });

    return { referenceId, userId, service, cost };
  }

  /** Ledger-only — balance was already deducted at reserve time. Non-fatal
   *  on failure: the AI call already succeeded, so we log for manual
   *  reconciliation rather than fail the response over it. */
  async commit(reservation: CreditReservation): Promise<void> {
    try {
      await this.prisma.aiCreditTransaction.create({
        data: {
          userId: reservation.userId,
          service: reservation.service,
          referenceId: reservation.referenceId,
          amount: 0,
          type: 'commit',
        },
      });
    } catch (e) {
      this.logger.error(
        `Credit commit failed to record (job succeeded regardless) user=${reservation.userId} ref=${reservation.referenceId}: ${e}`,
      );
    }
  }

  /** Refund a reservation when the handler throws after credits were
   *  already taken. */
  async refund(reservation: CreditReservation): Promise<void> {
    try {
      const updated = await this.prisma.userCredits.update({
        where: { userId: reservation.userId },
        data: { creditsBalance: { increment: reservation.cost } },
      });
      await this.prisma.aiCreditTransaction.create({
        data: {
          userId: reservation.userId,
          service: reservation.service,
          referenceId: reservation.referenceId,
          amount: reservation.cost,
          type: 'refund',
          balanceAfter: updated.creditsBalance,
        },
      });
    } catch (e) {
      this.logger.error(
        `Credit refund failed user=${reservation.userId} ref=${reservation.referenceId}: ${e}`,
      );
    }
  }
}
